### PR TITLE
Feature: SidecarSet injection supports Partial strategy

### LINKED ISSUE #1833: [feature request] SidecarSet support 'PartitionBased' policy to inject old/new revision according to Partition
**What would you like to be added**:

**Why is this needed**:

### PR DESCRIPTION
**Pull request by AiRanthem**
### Ⅰ. Describe what this PR does

Now, SidecarSet supports two policies in `injectionStrategy`:

1. `Always`, which is the default and current policy. Under this policy, all newly created Pods will be injected with the specified revision in `injectionStrategy`.
2. `Partial`, under which only when the Pod is **not** selected by the Selector explicitly configured in `UpdateStrategy` will it be definitely injected with the specified revision. In all other conditions, newly created Pods have a probability of being injected with the latest Sidecar, where the probability is `1 - UpdateStrategy.Partition`. If `Partition` is not a percentage or is not configured, its value is considered to be 0%.

### Ⅱ. Does this pull request fix one issue?

fixes #1833

### Ⅲ. Describe how to verify it
see the unit tests added

### CODE REVIEW

**Comment by AiRanthem** [comment]
@furykerry please help me to name the TODOSidecarSetInjectRevisionPolicy

File: pkg/webhook/sidecarset/validating/sidecarset_create_update_handler.go [hunk 1]
```diff
@@ -204,9 +204,10 @@ func (h *SidecarSetCreateUpdateHandler) validateSidecarSetInjectionStrategy(obj
 		}
 
 		switch revisionInfo.Policy {
-		case "", appsv1alpha1.AlwaysSidecarSetInjectRevisionPolicy:
+		case "", appsv1alpha1.AlwaysSidecarSetInjectRevisionPolicy, appsv1alpha1.PartialSidecarSetInjectRevisionPolicy:
 		default:
-			errList = append(errList, field.Invalid(field.NewPath("revision").Child("policy"), revisionInfo, fmt.Sprintf("Invalid policy %v, only support 'Always' currently", revisionInfo.Policy)))
+			errList = append(errList, field.Invalid(field.NewPath("revision").Child("policy"), revisionInfo, fmt.Sprintf("Invalid policy %v, supported: [%s, %s]",
+				revisionInfo.Policy, appsv1alpha1.AlwaysSidecarSetInjectRevisionPolicy, appsv1alpha1.PartialSidecarSetInjectRevisionPolicy)))
```
Comment (furykerry) [inline pkg/webhook/sidecarset/validating/sidecarset_create_update_handler.go:210]: consider add validation for undefined condition for partial strategy
1. setting updateStrategy selector and partition together 
2.  setting updateStrategy partition to integer values

File: pkg/util/tools.go [hunk 2]
```diff
@@ -227,6 +227,20 @@ func GetScaledValueFromIntOrPercent(intOrPercent *intstrutil.IntOrString, total
 	return 0, fmt.Errorf("invalid type: neither int nor percentage")
 }
 
+// ParsePercentageAsFloat64 parses a string as a percentage and returns the value as a float64.
+func ParsePercentageAsFloat64(s string) (float64, error) {
```
[the remark this answers is not in the document]
Comment (AiRanthem) [reply pkg/util/tools.go:231]: all I need is just to convert 14% to 0.14, CalculatePartitionReplicas is too heavy

[the remark this answers is not in the document]
Comment (furykerry) [reply pkg/util/tools.go:231]: then use the GetScaledValueFromIntOrPercent

[the remark this answers is not in the document]
Comment (AiRanthem) [reply pkg/util/tools.go:231]: In fact, ParsePercentageAsFloat64 is GetScaledValueFromIntOrPercent without the total and roundUp functionalities. Should we consider making GetScaledValueFromIntOrPercent use ParsePercentageAsFloat64?

File: pkg/webhook/sidecarset/validating/sidecarset_create_update_handler.go (hunk 1 above)
Comment (AiRanthem) [reply pkg/webhook/sidecarset/validating/sidecarset_create_update_handler.go:210]: setting updateStrategy selector and partition together is not necessary for:
1. if no selector set, all pods will have the chance to be injected the latest revision
2. if no partition set or it's not an integer value (which means partition only works for existing replicas, not for newly created ones), it is considered as 0 in web hook, meaning the chance to inject latest revision is 100%

File: pkg/util/tools.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (furykerry) [reply pkg/util/tools.go:231]: it is possible but not necessary, GetScaledValueFromIntOrPercent is already widely used in the code base and atomic enough

**Pull request merged**